src/global/common.rs

```diff
@@ -185,6 +187,138 @@ pub(crate) async fn find_package_records(conda_meta: &Path) -> miette::Result<Ve
     Ok(records)
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[must_use]
+pub(crate) enum StateChange {
+    AddedExposed(String, EnvironmentName),
+    RemovedExposed(String, EnvironmentName),
+    UpdatedExposed(String, EnvironmentName),
+    AddedPackage(PackageRecord, EnvironmentName),
+    AddedEnvironment(EnvironmentName),
+    RemovedEnvironment(EnvironmentName),
+}
+
+impl StateChange {
+    fn env(&self) -> &EnvironmentName {
+        match self {
+            StateChange::AddedExposed(_, env)
+            | StateChange::RemovedExposed(_, env)
+            | StateChange::UpdatedExposed(_, env)
+            | StateChange::AddedPackage(_, env)
+            | StateChange::AddedEnvironment(env)
+            | StateChange::RemovedEnvironment(env) => env,
+        }
+    }
+}
+
+impl fmt::Display for StateChange {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            StateChange::AddedExposed(exposed, env) => {
+                write!(
+                    f,
+                    "{}Added exposed '{exposed}' for environment '{env}'.",
+                    console::style(console::Emoji("✔ ", "")).green(),
+                )
+            }
+            StateChange::RemovedExposed(exposed, env) => {
+                write!(
+                    f,
+                    "{}Removed exposed '{exposed}' for environment '{env}'.",
+                    console::style(console::Emoji("✔ ", "")).green(),
+                )
+            }
+            StateChange::UpdatedExposed(exposed, env) => {
+                write!(
+                    f,
+                    "{}Updated exposed '{exposed}' for environment '{env}'.",
+                    console::style(console::Emoji("✔ ", "")).green(),
+                )
+            }
+            StateChange::AddedPackage(pkg, env) => write!(
+                f,
+                "{}Added package '{}={}' to environment '{env}'.",
+                console::style(console::Emoji("✔ ", "")).green(),
+                pkg.name.as_normalized(),
+                pkg.version,
+            ),
+            StateChange::AddedEnvironment(env) => write!(
+                f,
+                "{}Added environment '{env}'.",
+                console::style(console::Emoji("✔ ", "")).green()
+            ),
+            StateChange::RemovedEnvironment(env) => write!(
+                f,
+                "{}Removed environment '{env}'.",
+                console::style(console::Emoji("✔ ", "")).green()
+            ),
+        }
+    }
+}
+
+#[must_use]
+#[derive(Debug, Default)]
+pub(crate) struct StateChanges {
+    changes: Vec<StateChange>,
+    has_updated: bool,
+}
```

Comment:
This is the interesting bit. By returning this struct we can pick and choose what to report. For example, we never want to report changes done by reverting.